Handle PostConstruct and PreDestroy Exceptions in LifeCycleManager

pettyjamesm opened the pull request:
Handle catch and log exceptions thrown from `@PreDestroy` methods in`LifeCycleManager::stop()` so that other stop behaviors have a chance to run.

pettyjamesm commented:
Test failures seem unrelated and likely due to a racy assertion, tests pass locally for me. This implementation chooses to swallow exceptions raised by `@PreDestroy` methods and just log them, but they could be re-thrown at the end of `LifeCycleManager::stop()` as well. If that's preferred, here's a diff of that implementation: https://github.com/airlift/airlift/compare/master...pettyjamesm:handle-predestroy-exceptions-and-rethrow

findepi commented on bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleStopException.java:
```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright 2010 Proofpoint, Inc.
```
?
If i read `src/license/LICENSE-HEADER.txt` correctly, this line is not mandatory.

(same in other new files)

electrum replied:
Yes, we shouldn't add this copyright line to new files, since it isn't correct

pettyjamesm replied:
Sure, will remove the header from new files.

findepi commented on bootstrap/src/test/java/io/airlift/bootstrap/DestroyExceptionInstance.java:
```diff
@@ -0,0 +1,48 @@
+    @PostConstruct
+    public void postDestroyException()
+    {
+        TestLifeCycleManager.note("postDestroyException");
```
- `postDestroyException` -> `post*Construct*...` ?
- did you intent to have `throw ` here? Is this method needed at all?

pettyjamesm replied:
Not strictly, was just following the convention of other instances. I'll remove.

findepi commented on bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleManager.java:
```diff
@@ -133,17 +133,31 @@ public void stop()
+                }
+                catch (Exception e) {
+                    Throwable t = (e instanceof InvocationTargetException) ? ((InvocationTargetException) e).getTargetException() : e;
+                    log.error("Exception in PreDestroy method %s::%s()", obj.getClass().getName(), preDestroy.getName(), t);
```
Is `t` being logged? (fmt string has 2 `%s`s only)

If you intent to log the stacktrace in full, `t` should be the first param... but I don't think it's a good idea, since `t` gets propagated eventually.

findepi replied:
@electrum if there are multiple components failing, logging one line immediately can improve log output readability. 
IMO it's better to log here, but just `t` `toString()` instead of full stacktrace.

dain replied:
The general rule is "log or throw, not both".

pettyjamesm replied:
Re: the first comment, I didn't look hard enough at the logger API used here. Throwable should indeed be the first argument if being logged. Will do that.

Regarding the log ^ throw debate, I'm torn about how to proceed. If the lifecycle is stopping because the JVM is exiting, it definitely makes sense to log errors immediately because the logger infrastructure may be shutting down or the JVM may force-exit after some timeout. Logging immediately should be clearly preferable.

However, lifecylce stops initiated by an application that will continue to run will probably prefer to handle the exceptions all at once, probably via a checked exception type so there's only one type of exception to handle the problematic (or "dirty") shutdown.

These two use cases seem at odds with one another, so we can either:
- pick one at the detriment of the other use case
- break the log ^ throw convention in this particular case
- make these two use cases call distinct methods and support both modes.

I'm going to pursue the two method approach and update the CR, but I'm happy to implement either of the other two options if you guys prefer. Let me know.

electrum commented on bootstrap/src/test/java/io/airlift/bootstrap/TestLifeCycleManager.java:
```diff
@@ -212,6 +214,44 @@ public void testJITInjection()
+            Set<String> suppressedMessages = Arrays.stream(e.getSuppressed())
+                    .map(Throwable::getMessage)
+                    .collect(Collectors.toSet());
+            assertEquals(suppressedMessages, Sets.newHashSet("preDestroyExceptionOne", "preDestroyExceptionTwo"));
```
Nit: use `ImmutableSet.of`

pettyjamesm replied:
Easy fix, will change.

electrum commented on bootstrap/src/test/java/io/airlift/bootstrap/TestLifeCycleManager.java:
```diff
@@ -212,6 +214,44 @@ public void testJITInjection()
+            fail("Expected LifeCycleStopException to be throw, found: " + e.getClass());
+        }
+
+        assertEquals(new HashSet<>(stateLog), Sets.newHashSet(
```
`ImmutableSet.copyOf` and `ImmutableSet.of`

pettyjamesm replied:
Will change.

electrum commented on bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleManager.java:
```diff
@@ -133,17 +133,31 @@ public void stop()
+                    preDestroy.invoke(obj);
+                }
+                catch (Exception e) {
+                    Throwable t = (e instanceof InvocationTargetException) ? ((InvocationTargetException) e).getTargetException() : e;
```
This can be simplified to
```java
Throwable t = (e instanceof InvocationTargetException) ? e.getCause() : e;
```
Per the documentation, `getCause()` is the preferred way.

pettyjamesm replied:
Easy fix, will change.

pettyjamesm commented:
I've added a new commit to put the exception handling mechanism during `stop()` behind an interface so that log / suppressing exceptions can be done in whatever mechanism makes sense. This also enables calling `@PreDestroy` methods on instances that fail during their `@PostConstruct` phase, which I added as well.

findepi commented:
Re: "log ^ throw" -- I already expressed my opinion in https://github.com/airlift/airlift/pull/739#discussion_r300422597
I am deferring to @electrum and @dain on a decision.

pettyjamesm commented:
@findepi any feedback / opinions on the new commit added? I was trying to strike a balance for how to handle that "log ^ throw" problem and saw the opportunity to clean up instances that fail to `@PostConstruct` cleanly

findepi commented:
@pettyjamesm from engineering perspective this is well done. I am not convinced yet, however, that we need this flexibility.
I would prefer to settle on one simple option and revisit the decision when some real-life situation demands this.

pettyjamesm commented:
@findepi still waiting on a review follow up from @dain or @electrum, but regarding your concern about it being over complicated- I both agree and unfortunately don't see a way to accomplish this more simply. Ignoring the fact that this enables "arbitrary" implementations of `LifeCycleStopFailureHandler` to be provided by the caller, we have 3 use cases internal to this class alone:
1. Stop and log errors immediately, throw a `LifeCycleStopException` at the end if any occurred but without any reference to the already logged exception(s)
1. Stop and attach all errors as suppressed to a parent `LifeCylceStopException` without logging
1. Stop a single instance that failed to fully `@PostConstruct` successfully and attach all exceptions as suppressed to the exception raised from start-up logic.

I played around trying to find a simpler way to handle the cases mentioned above without resorting to a full blown interface type, but was ultimately unsuccessful. Whether we choose to make the `LifeCycleStopFailureHandler` interface and `stop(LifeCycleStopFailureHandler)` method public is another matter- I'll gladly make them private if that's preferable.

electrum commented on bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleManager.java:
```diff
@@ -110,14 +110,52 @@ public void start()
+    }
+
+    /**
+     * Stop the life cycle - all instances will have their {@link javax.annotation.PreDestroy} method(s) called
```
`PreDestroy` probably doesn't need to be qualified, since it's already imported. Same for other usages.

electrum commented on bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleStopException.java:
```diff
@@ -0,0 +1,10 @@
+package io.airlift.bootstrap;
+
+public class LifeCycleStopException
+        extends Exception
```
Let's make this extend `RuntimeException`. No point in forcing the caller to handle it, since they don't have any reasonable way to do so. I did a quick survey of the ~20 instances in Presto and they all either log or simply rethrow it.

electrum commented on bootstrap/src/test/java/io/airlift/bootstrap/DestroyExceptionInstance.java:
```diff
@@ -0,0 +1,26 @@
+    @Inject
+    public DestroyExceptionInstance(DependentInstance ignored)
+    {
+    }
```
Let's add a comment
```java
{
    // the constructor exists to force injection of the dependent instance
}

electrum commented on bootstrap/src/test/java/io/airlift/bootstrap/TestLifeCycleManager.java:
```diff
@@ -212,6 +215,76 @@ public void testJITInjection()
+                    .collect(Collectors.toSet());
+            assertEquals(ImmutableSet.copyOf(suppressedMessages), ImmutableSet.of("preDestroyExceptionOne", "preDestroyExceptionTwo"));
+        }
+        catch (Exception e) {
```
Let's remove this catch block. Better to let the exception propagate and fail the test, since that's the normal behavior, and it will provide the full stack trace of the unexpected exception.

electrum commented on bootstrap/src/test/java/io/airlift/bootstrap/TestLifeCycleManager.java:
```diff
@@ -212,6 +215,76 @@ public void testJITInjection()
+        catch (LifeCycleStopException e) {
+            assertEquals(e.getSuppressed().length, 0, "Suppressed exceptions list should be empty");
+        }
+        catch (Exception e) {
```
Same as above

electrum commented on bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleManager.java:
```diff
@@ -133,18 +171,33 @@ public void stop()
+                        log.error(t, "Exception in PreDestroy method %s::%s()", obj.getClass().getName(), preDestroy.getName());
+                    }
+                    else {
+                        log.debug(t, "Exception in PreDestroy method %s::%s()", obj.getClass().getName(), preDestroy.getName());
```
We shouldn't log here at all. Even at debug level, log-and-throw is still confusing.

electrum commented on bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleManager.java:
```diff
@@ -221,14 +238,49 @@ public void addInstance(Object instance)
+
+    /**
+     * Unwraps {@link InvocationTargetException} instances to their underlying cause, otherwise returns the provided
+     * Exception
```
Nit: we don't have a hard line length limit, so either put this all on one line, or balance the text onto two lines (having a single word on a line looks strange)

electrum commented on bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleStopException.java:
```diff
@@ -1,7 +1,7 @@
 
 public class LifeCycleStopException
-        extends Exception
+        extends RuntimeException
```
Move this change to the previous commit (where the class is introduced)

electrum commented on bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleManager.java:
```diff
@@ -40,6 +40,16 @@
+     * whatever way makes the most sense. Implementations should not raise exceptions of
+     * their own
+     */
+    public interface LifeCycleStopFailureHandler
```
Let's make this private for now. We can make it public later if there is a use case.

electrum commented on bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleManager.java:
```diff
@@ -119,13 +129,26 @@ public void start()
+            }
+        }
+
+        CollectForSuppressedExceptionsHandler handler = stop(new CollectForSuppressedExceptionsHandler());
```
We can simplify this with a lambda
```java
List<Exception> failures = new ArrayList<>();
stop((klass, method, exception) -> failures.add(exception));

electrum commented on bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleManager.java:
```diff
@@ -140,22 +163,34 @@ public void stopWithoutFailureLogging()
+            }
+        }
+
+        if (stop(new LogEachAndDiscardFailureHandler()).exceptionOccurred) {
```
Similarly, this seems simpler as a lambda:
```java
AtomicBoolean failure = new AtomicBoolean();
stop((klass, method, exception) -> {
    log.error(exception, "Exception in PreDestroy method %s::%s()", klass.getName(), method.getName());
    failure.set(true);
});
if (failure.get()) {
    throw new LifeCycleStopException();
}

electrum commented on bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleManager.java:
```diff
@@ -140,22 +163,34 @@ public void stopWithoutFailureLogging()
+     * @return The same {@link LifeCycleStopFailureHandler} instance passed to this method
      */
-    private List<Throwable> stopInternal(boolean logFailuresImmediately)
+    public <T extends LifeCycleStopFailureHandler> T stop(T handler)
```
After changing to lambda, we don't need this to return the handler -- it can be `void`. The caller already has the handler if they need it.

electrum commented on bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleManager.java:
```diff
@@ -221,14 +238,49 @@ public void addInstance(Object instance)
+            catch (Exception e) {
+                Exception cause = unwrapInvocationTargetException(e);
+                log.debug(cause, "Exception in PostConstruct method %s::%s()", obj.getClass().getName(), postConstruct.getName());
+                stopInstance(obj, (Class<?> klass, Method method, Exception exception) -> {
```
I don't think the lambda parameter types are needed here. It can be
```java
stopInstance(obj, (klass, method, exception) -> {

electrum commented on bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleManager.java:
```diff
@@ -221,14 +238,49 @@ public void addInstance(Object instance)
+            }
+            catch (Exception e) {
+                Exception cause = unwrapInvocationTargetException(e);
+                log.debug(cause, "Exception in PostConstruct method %s::%s()", obj.getClass().getName(), postConstruct.getName());
```
Rather than logging here and suffering from log-and-throw, what if we enhance the thrown exception:
```java
LifecycleStartException failure = new LifecycleStartException(
        format("Exception in PostConstruct method %s::%s()", obj.getClass().getName(), postConstruct.getName()),
        unwrapInvocationTargetException(e));
stopInstance(obj, (klass, method, exception) -> {
    String message = format("Exception in PreDestroy method %1$s::%2$s() after PostConstruct failure in %1$s::%3$s()", klass.getName(), method.getName(), postConstruct.getName());
    failure.addSuppressed(new RuntimeException(message, exception));
});
throw failure;
```
With this change, we'd introduce a new unchecked `LifecycleStartException` and change the `start()` method to throw that for all failures. This aligns it with `stop()` for the same reasons -- callers don't have any reasonable way to handle the underlying exception.

pettyjamesm commented:
@electrum thanks for the review, I've combined the two commits into one and addressed all your CR feedback.

electrum commented:
This is great! Thanks a lot for improving this and putting up with several rounds of review.
